### ISSUE TITLE
Agda Prelude

### ISSUE THREAD

**Issue by asr** [agda/agda#1875]
Stephen Diehl wrote in [The Joy and Agony of Haskell in Production](http://www.stephendiehl.com/posts/production.html):

\begin{quote}

**Consider rolling an internal prelude**. As we’ve all learned the hard way, the Prelude is not your friend. The consensus historically has favored the “Small Prelude Assumption” which presupposes that tools get pushed out into third party modules, even the core tools that are necessary to do anything (text, bytestring, vector, etc). This makes life easier for library authors at the cost of some struggle for downstream users.

In practice any non-trivial business logic module can very easily have 100+ lines just of imports, and frankly it gets tiring. One common way of abstracting this is by rolling a   custom prelude using module reexports. Consider a minimal use case like the following:

``` haskell
module MegaCorpPrelude ( 
  module Exports,
) where

import Data.Int as Exports
import Data.Tuple as Exports
import Data.Maybe as Exports
import Data.String as Exports
import Data.Foldable as Exports
import Data.Traversable as Exports

import Control.Monad.Trans.Except
  as Exports
  (ExceptT(ExceptT), Except, except, runExcept, runExceptT, 
   mapExcept, mapExceptT, withExcept, withExceptT)
```

This can be put into a cabal package which transitively pulls in the core dependencies and then is used in our downstream module.

``` haskell
{-# LANGUAGE NoImplicitPrelude #-}

import MegaCorpPrelude
```

\end{quote}

I like the idea of defining our own `Agda.Prelude` because
- this module will simplify us the support of three major versions of GHC (I'm not sure about UHC),
- in this module we could consistently implement that `Data.Utils.Foo` re-exports `Data.Foo` and
- we could avoid _many_ lines of just imports (as mentioned above).

As I see it, we could start to define/use `Agda.Prelude` peacefully. Once this module has all that is required and all the modules are importing this module, we will add the `NoImplicitPrelude` option.

Drawback: Since `Agda.Prelude` will be imported by all the modules, a change in this module generates a recompilation of all the modules.

What do you think?

**Comment by UlfNorell** [comment]
> this module will simplify us the support of three major versions of GHC (I'm not sure about UHC)

We're not pretending that you will ever be able to compile Agda with UHC are we?

As to the actual proposal, I don't think it will save many lines of imports. Of 310 source modules only 40 have more than 10 non-Agda imports and 150 have less than 5. The most imports are 23 in `Agda.TypeChecking.Monad.Base`. It could still be worth it, I suppose, just to avoid `Control.Applicative` and `Data.Foldable` issues, but I don't really feel that those are a huge problem.

**Comment by asr** [comment]
> We're not pretending that you will ever be able to compile Agda with UHC are we?

No. I was thinking about that versions of UHC supports the UHC backend. Right now the backend supports UHC 1.1.9.2 and 1.1.9.3.

**Comment by UlfNorell** [comment]
Right, of course.

**Comment by asr** [comment]
> As to the actual proposal, I don't think it will save many lines of imports. Of 310 source modules only 40 have more than 10 non-Agda imports and 150 have less than 5. The most imports are 23 in Agda.TypeChecking.Monad.Base. It could still be worth it, I suppose, just to avoid Control.Applicative and Data.Foldable issues, but I don't really feel that those are a huge problem.

I'm also counting various of the Agda.Util.Foo imports.

**Comment by phile314** [comment]
The UHC backend shouldn't be relevant to this discussion. For the time being, we (officially) want to support only one UHC version, everything else is too much work.

The whole UHC backend cpp magic is only in ~1 file ( https://github.com/agda/agda/blob/master/src/full/Agda/Compiler/UHC/Bridge.hs ), so there should also be no cpp-related problems.

---

Simplifying our GHC compatibility related code would be great!

**Comment by andreasabel** [comment]
In principle I like the idea to bundle common imports in a Prelude.  This would save us fiddling with the import statements a bit.  `Agda.TypeChecking.Monad` is such a prelude for the type checking modules.
On the downside, frequent changes to the Prelude would require frequent long compilation times.

**Comment by UlfNorell** [comment]
Hopefully it would be reasonably stable after a short while of figuring out what should go in it.

**Comment by UlfNorell** [comment]
I know we talked about this at some point, but I don't see it mentioned here, so:

Note that the Haskell module system doesn't (afaik) let us capture one of the more common import patterns:
```haskell
import Data.Map (Map)
import qualified Data.Map as Map
```
(and `Set` and `HashMap` and `IntSet` and ...)

**Comment by jespercockx** [comment]
> I know we talked about this at some point, but I don't see it mentioned here, so:
> 
> Note that the Haskell module system doesn't (afaik) let us capture one of the more common import patterns:
> 
> ```haskell
> import Data.Map (Map)
> import qualified Data.Map as Map
> ```
> 
> (and `Set` and `HashMap` and `IntSet` and ...)

The `mono-traversable` library provides type classes `IsMap` and `IsSet` (and much more stuff) that unify the interface to these standard datastructures so they do not need to be qualified (see https://hackage.haskell.org/package/mono-traversable-1.0.15.1/docs/Data-Containers.html). Maybe we could consider adding this as a dependency so we can avoid these qualified imports altogether? It also has other nice stuff like typeclasses for monomorphic datastructures such as `IntSet` that we use a lot (and currently we need to qualify all the time).

**Comment by nad** [comment]
The page you link to includes the following text: "Warning: This module should be considered highly experimental".
